src/engine/AbsoluteValueConstraint.cpp

```diff
@@ -0,0 +1,454 @@
+
+    Equation negativeEquation( Equation::EQ );
+    negativeEquation.addAddend( 1, _b );
+    negativeEquation.addAddend( 1, _f );
+    negativeEquation.setScalar( 0 );
+    negativePhase.addEquation( negativeEquation );
+
+    return negativePhase;
+}
+
+PiecewiseLinearCaseSplit AbsoluteValueConstraint::getPositiveSplit() const
+{
+    PiecewiseLinearCaseSplit positivePhase;
+
+    // Positive phase: b >= 0, b - f = 0
+    positivePhase.storeBoundTightening( Tightening( _b, 0.0, Tightening::LB ) );
+
+    //b - f = 0
+    Equation positiveEquation( Equation::EQ );
+    positiveEquation.addAddend( 1, _b );
+    positiveEquation.addAddend( -1, _f );
+    positiveEquation.setScalar( 0 );
+    positivePhase.addEquation( positiveEquation );
+
+    return positivePhase;
+}
+
+bool AbsoluteValueConstraint::phaseFixed() const
+{
+    return _phaseStatus != PhaseStatus::PHASE_NOT_FIXED;
+}
+
+PiecewiseLinearCaseSplit AbsoluteValueConstraint::getValidCaseSplit() const
+{
+    ASSERT( _phaseStatus != PhaseStatus::PHASE_NOT_FIXED );
+
+    if ( _phaseStatus == PhaseStatus::PHASE_POSITIVE )
+        return getPositiveSplit();
+
+    return getNegativeSplit();
+}
+
+void AbsoluteValueConstraint::eliminateVariable( unsigned /* variable */, double /* fixedValue */ )
+{
+    // In an absolute value constraint, if a variable is removed the
+    // entire constraint can be discarded
+    _haveEliminatedVariables = true;
+}
+
+void AbsoluteValueConstraint::updateVariableIndex( unsigned oldIndex, unsigned newIndex )
+{
+    ASSERT( oldIndex == _b || oldIndex == _f );
+    ASSERT( !_assignment.exists( newIndex ) &&
+            !_lowerBounds.exists( newIndex ) &&
+            !_upperBounds.exists( newIndex ) &&
+            newIndex != _b && newIndex != _f );
+
+    if ( _assignment.exists( oldIndex ) )
+    {
+        _assignment[newIndex] = _assignment.get( oldIndex );
+        _assignment.erase( oldIndex );
+    }
+
+    if ( _lowerBounds.exists( oldIndex ) )
+    {
+        _lowerBounds[newIndex] = _lowerBounds.get( oldIndex );
+        _lowerBounds.erase( oldIndex );
+    }
+
+    if ( _upperBounds.exists( oldIndex ) )
+    {
+        _upperBounds[newIndex] = _upperBounds.get( oldIndex );
+        _upperBounds.erase( oldIndex );
+    }
+
+    if ( oldIndex == _b )
+        _b = newIndex;
+    else
+        _f = newIndex;
+}
+
+bool AbsoluteValueConstraint::constraintObsolete() const
+{
+    return _haveEliminatedVariables;
+}
+
+void AbsoluteValueConstraint::getEntailedTightenings( List<Tightening> &tightenings ) const
+{
+    ASSERT( _lowerBounds.exists( _b ) && _lowerBounds.exists( _f ) &&
+            _upperBounds.exists( _b ) && _upperBounds.exists( _f ) );
+
+    // Upper bounds
+    double bUpperBound = _upperBounds[_b];
+    double fUpperBound = _upperBounds[_f];
+    // Lower bounds
+    double bLowerBound = _lowerBounds[_b];
+    double fLowerBound = _lowerBounds[_f];
+
+    // F's lower bound should always be non-negative
```

Comment:
Second place where we return this tightening, I believe.

Reply:
Yes, there are two mechanisms for getting tighter bounds: through the "notify" functions, where the constraint initiates sending the information; or through the getEntailedTightengings function, where the caller initiates the process.